Pass all graphdriver create() parameters in a struct

rhvgoyal opened the pull request:
This allows for easy extension of adding more parameters to existing
parameters list. Otherwise adding a single parameter changes code
at so many places.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
cc @crosbymichael @thaJeztah

thaJeztah commented:
is this going to affect the experimental graphdriver-plugins? https://github.com/docker/docker/blob/master/experimental/plugins_graphdriver.md

thaJeztah commented:
/cc @cpuguy83

rhvgoyal commented:
@thaJeztah I don't think that graphdriver plugins will see any affect. All the changes are internal.

dmcgowan commented:
Having the non-optional parameters outside of the options struct is a much cleaner pattern. I think this would include both the id and the parent and would end up looking more like...

```
func (a *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) error
```

What is nice about that approach is in most trivial cases nil could be passed in for the opts. The way it is coded up now a pointer is passed in but passing in nil is never allowed. Since that would be a large update I will let other maintainers weigh in before asking to update :)

rhvgoyal commented:
`parent` is optional parameter as well.

crosbymichael commented:
Moving the id out like @dmcgowan said makes sense to me.  If parent is optional then it would be best to leave in the opts struct

dmcgowan commented:
I am fine putting parent in options struct. Parent can be seen as optional or `""` can be viewed as scratch, doesn't matter. Was just suggesting having it outside since it is commonly given throughout the tests while the other options will usually be empty.

rhvgoyal commented:
Ok, I am fine either way. Please decide if you want `parent` inside struct or outside. For now, I will start with putting it outside struct. Do let me know if you don't like it.

crosbymichael commented:
@rhvgoyal that is fine with me

dmcgowan commented:
> For now, I will start with putting it outside struct.

:+1:

rhvgoyal commented:
Moved `id` and `parent` fields outside `CreateOpts` struct.

dmcgowan commented on daemon/graphdriver/driver.go:
```diff
@@ -50,10 +57,10 @@ type ProtoDriver interface {
-	CreateReadWrite(id, parent, mountLabel string, storageOpt map[string]string) error
+	CreateReadWrite(id, parent string, opts *CreateOpts) error
 	// Create creates a new, empty, filesystem layer with the
 	// specified id and parent and mountLabel. Parent and mountLabel may be "".
```
Update comment to mention create options instead of mount label and mention that the create options may be nil.

dmcgowan commented:
Code looks good, just one comment needs updating

crosbymichael commented:
LGTM

rhvgoyal committed:
Pass all graphdriver create() parameters in a struct

This allows for easy extension of adding more parameters to existing
parameters list. Otherwise adding a single parameter changes code
at so many places.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@dmcgowan Updated patch to address the comment. PTAL.

dmcgowan commented:
LGTM

thaJeztah reviewed (approved):
if it looks good to @dmcgowan, it LGTM
